// interpret_and_control/interpret_and_control.js


var Client = require('ibmiotf');

function interpret_and_control(mqttCreds, cloudant){
    this.cloudant = cloudant;
    // set up mqtt_handler
    var mqttConfig = {
       "org" : mqttCreds.org,
       "id" : mqttCreds.id,
       "auth-key" : mqttCreds.apiKey,
       "auth-token" : mqttCreds.apiToken
    }
    this.client = new Client.IotfApplication(mqttConfig);
    this.client.connect();

    this.client.on("error", function (err) {
        console.log("Error : " + err);
    });

};

interpret_and_control.prototype.connectionStatus = function(){
    return this.client.getOrganizationDetails();
};

interpret_and_control.prototype.imageKeywords = function(keywords, time, location){
    var identified = false;


    for (var keyword in keywords){
        switch (keyword) {
            case "Fire":
                this.sendCommand("pi", "dronepi", "movement", "json", "UP");
                this.sendCommand("Android", "phone", "imageRecognition", "json", keyword);
                var data = {d: {text:"fire"}};
                this.sendCommand("Android", "phone", "imageRecognition", "json", JSON.stringify(data));
                var color = {d: {r:155, g:255, b:0, alpha:120}};
                this.sendCommand("Android", "phone", "color", "json", JSON.stringify(color));
                var alert = {d: {text:"BANG"}};
                this.sendCommand("Android", "phone", "alert", "json", JSON.stringify(alert));
                this.logEvent(keyword, time, location);
                identified = true;
                break;
            case "Person":
                this.sendCommand("pi", "dronepi", "movement", "json", "STOP");
                this.sendCommand("Android", "phone", "imageRecognition", "json", "Person");
                this.logEvent(keyword, time, location);
                identified = true;
            default:

        }
    }
    return identified;
}

interpret_and_control.prototype.speechTranscript = function(transcript){
    // if exists do shit?

    if(!transcript.equals("")){
        this.sendCommand("pi", "dronepi", "movement", "json", "STOP");
        this.sendCommand("Android", "phone", "imageRecognition", "json", "Person");
        //somehow to tone Analysis?
    }
}

interpret_and_control.prototype.sendCommand = function(deviceType, deviceId, commandType, format, data){
    var client = this.client;
    client.on("connect", function(){
        client.publishDeviceCommand(deviceType, deviceId, commandType, format, data);
    });
}

interpret_and_control.prototype.sendEvent = function(deviceType, deviceId, commandType, format, data){
    var client = this.client;
    client.on("connect", function(){
        client.publishDeviceEvent(deviceType, deviceId, commandType, format, data);
    });
}

interpret_and_control.prototype.logEvent = function(type, time, location){
    var eventLog = this.cloudant.use('eventlog');
    var docID = time;
    var data = {
        eventType : type,
        eventLocation: location,
    }
    eventLog.insert(docID, data, function(err, body){
        if(err){
            console.log("Error inserting event " + type + "from time:" + time);
        }
    });
}


module.exports = interpret_and_control;
